core-api/src/main/java/com/optimizely/ab/odp/ODPManager.java

```diff
@@ -58,4 +65,89 @@ public Boolean updateSettings(String apiHost, String apiKey, List<String> allSeg
     public void close() {
         eventManager.stop();
     }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private ODPSegmentManager segmentManager;
+        private ODPEventManager eventManager;
+        private ODPApiManager apiManager;
+        private Integer cacheSize;
+        private Integer cacheTimeoutSeconds;
+        private Integer batchSize;
+        private Integer flushIntervalMillis;
+        private Cache<List<String>> cacheImpl;
+
+        public Builder withApiManager(ODPApiManager apiManager) {
+            this.apiManager = apiManager;
+            return this;
+        }
+
+        public Builder withSegmentManager(ODPSegmentManager segmentManager) {
+            this.segmentManager = segmentManager;
+            return this;
+        }
+
+        public Builder withEventManager(ODPEventManager eventManager) {
+            this.eventManager = eventManager;
+            return this;
+        }
+
+        public Builder withSegmentCacheSize(Integer cacheSize) {
+            this.cacheSize = cacheSize;
+            return this;
+        }
+
+        public Builder withSegmentCacheTimeout(Integer cacheTimeoutSeconds) {
+            this.cacheTimeoutSeconds = cacheTimeoutSeconds;
+            return this;
+        }
+
+        public Builder withSegmentCache(Cache<List<String>> cacheImpl) {
+            this.cacheImpl = cacheImpl;
+            return this;
+        }
+
+        public Builder withEventBatchSize(Integer batchSize) {
+            this.batchSize = batchSize;
+            return this;
+        }
+
+        public Builder withEventFlushInterval(Integer flushIntervalMillis) {
+            this.flushIntervalMillis = flushIntervalMillis;
+            return this;
+        }
+
+        public ODPManager build() {
+            if ((segmentManager == null || eventManager == null) && apiManager == null) {
+                logger.warn("ApiManager instance is needed when using default EventManager or SegmentManager");
```

Comment:
I wanted to but the problem is, APIManager contains the http client implementation which is not part of the core sdk. This is inside the core SDK. Core SDK cannot call and initialize the DefaultApiManager because that is in the `http-client-impl`. I can add a no-op or logging ApiManager as a default but that would be useless because that will not be making any external calls. What do you suggest?

Reply:
@zashraf1985 I see. The downside is that the client always have to set ODPManager even for default settings, but I see this is consistent with the current java-sdk patterns.

Reply:
Yes the way our java sdk is designed, if you do not use factory, then you have to build event processor, datafile manager and odp api manager using their builders and then pass on to the optimizely instance. I have tried to follow the same pattern here. If a user wants default settings, they can use the OptimizelyFactory which will automatically use the DefaultODPApiManager among other things.